File: Python3/Boats to Save People.py
#1
class Solution:
    def numRescueBoats(self, people: List[int], limit: int) -> int:
        people.sort(reverse=True)
        i, j = 0, len(people) - 1
        while i <= j:
            if people[i] + people[j] <= limit: j -= 1
            i += 1
        return i

#2
class Solution:
    def numRescueBoats(self, people: List[int], limit: int) -> int:
        people.sort()
        left = 0
        right = len(people) - 1
        boat = 0
        while left<=right:
            if people[left]+people[right]<=limit:
                boat += 1
                right -= 1
                left += 1
            else:
                boat += 1
                right -= 1

        return boat

File: Golang/Boats to Save People.go
#1
func numRescueBoats(people []int, limit int) int {
    sort.Ints(people)
    start := 0
    end := len(people) - 1
    res := 0
    for start <= end {
       if people[start] + people[end] <= limit {
           start++
           end--
           res++
       } else {
           end--
           res++
       }
    }
    return res
}

#2
func numRescueBoats(people []int, limit int) int {
    sort.Slice(people, func(i, j int) bool {
        return people[i] > people[j]
    })
    start := 0
    end := len(people) - 1
    for start <= end {
       if people[start] + people[end] <= limit {
           end--
       }
       start++
    }
    return start
}